include/envoy/upstream/cluster_manager.h

```diff
@@ -209,6 +209,11 @@ class ClusterManager {
   addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
 
   virtual ClusterManagerFactory& clusterManagerFactory() PURE;
+
+  /**
+   * @return TimeSource& the time-source used with the cluster manager.
+   */
+  virtual TimeSource& timeSource() PURE;
```

Comment:
Is time really an essential property of `ClusterManager`?

Reply:
Offhand, I am not an expert in the class; this is the first I've touched it, IIRC.

However, it appears based on its API to care about load-balancing, which I think typically has a time-component implied, no?

From a practical perspective, I need to find a time-source from places that already have a ClusterManager plumbed in and not much else.  E.g. AsyncClientImpl

IMO you don't need a lot of excuses to want to know about time. Even a pure algorithm may want to employ timeouts, etc. Something I'm not proud of here is that PerfAnnotation in this PR still needs its own prod time-source, but I can't solve all problems in this PR, and I'm not sure we have a TimeSource available everywhere I might want to measure perf.

Reply:
It is practical but will enlarge the CL significantly from what I can tell.  Can I suggest a compromise where I add a TODO to the interfaces where it doesn't make (IMO) obvious sense (Dispatcher, arguably Server) to remove them when we've got everything plumbed directly?

I think the scary one is filters; I didn't want to change how every filter gets instantiated, preferring instead to provide access to the TimeSource via FilterConfig.  Happy to remove that, but it will require changing every filter.